File: 2807.insert-greatest-common-divisors-in-linked-list.py
#
# @lc app=leetcode id=2807 lang=python3
#
# [2807] Insert Greatest Common Divisors in Linked List
#

# @lc code=start
from math import gcd
from typing import Optional, List

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def insertGreatestCommonDivisors(self, head: Optional[ListNode]) -> Optional[ListNode]:
        current = head
        while current and current.next:
            # Get next node for calculation of gcd and insertion point.
            next_node = current.next
            # Calculate gcd of current value and next value.
            gcd_value = gcd(current.val, next_node.val)
            # Create a new node with this gcd value.
            gcd_node = ListNode(gcd_value) 
            # Link new node between current and next node. 
            current.next = gcd_node 
            gcd_node.next = next_node 
            # Move to next original node in sequence (skip over inserted node). 
            current = next_node 
        return head 
# @lc code=end

File: 2807.insert-greatest-common-divisors-in-linked-list.java
#
# @lc app=leetcode id=2807 lang=java
#
# [2807] Insert Greatest Common Divisors in Linked List
#
# @lc code=start
/**
* Definition for singly-linked list.
* public class ListNode {
*     int val;
*     ListNode next;
*     ListNode() {}
*     ListNode(int val) { this.val = val; }
*     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
* }
*/
import java.util.*; 
class Solution {
    public ListNode insertGreatestCommonDivisors(ListNode head) { 
        if (head == null || head.next == null) return head; 
        ListNode current = head; 
        while (current != null && current.next != null) { 
            int gcdValue = gcd(current.val, current.next.val); 
            ListNode gcdNode = new ListNode(gcdValue); 
            gcdNode.next = current.next; 
            current.next = gcdNode; 
            current = gcdNode.next; 
        } 
        return head; 
    } 
    private int gcd(int a, int b) { 
        if (b == 0) return a; 
        return gcd(b, a % b); 
    } 
}
# @lc code=end